### PR TITLE
HBASE-23331: Documentation for HBASE-18095

### PR DESCRIPTION
**Pull request by bharathv**

### CODE REVIEW

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 31s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
||| _ HBASE-18095/client-locate-meta-no-zookeeper Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   5m 32s |  HBASE-18095/client-locate-meta-no-zookeeper passed  |
| +0 :ok: |  refguide  |   4m 46s |  branch has no errors when building the reference guide. See footer for rendered docs, which you should manually inspect.  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   5m  4s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +0 :ok: |  refguide  |   4m 50s |  patch has no errors when building the reference guide. See footer for rendered docs, which you should manually inspect.  |
||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 19s |  The patch does not generate ASF License warnings.  |
|  |   |  22m 12s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.5 Server=19.03.5 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1164/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/1164 |
| JIRA Issue | HBASE-23331 |
| Optional Tests | dupname asflicense refguide |
| uname | Linux 40c0e195bda9 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/Base-PreCommit-GitHub-PR_PR-1164/out/precommit/personality/provided.sh |
| git revision | HBASE-18095/client-locate-meta-no-zookeeper / 2eead6c9ff |
| refguide | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1164/1/artifact/out/branch-site/book.html |
| refguide | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1164/1/artifact/out/patch-site/book.html |
| Max. process+thread count | 96 (vs. ulimit of 10000) |
| modules | C: . U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1164/1/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

File: dev-support/design-docs/HBASE-18095-Zookeeper-less-client-connection-design.adoc [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+* Meta table region locations
+
+ZK is deemed the source of truth since other processes that maintain the cluster state persist the changes to this data into ZK. So it is an obvious place to look at for clients to fetch the latest cluster state.  However this comes with it’s own set of problems, some of them are below.
+
+* Timeouts and retry logic for ZK clients are managed separately from HBase configuration. This is more administration overhead for end users (example: multiple timeouts are to be configured for different types of RPCs. client->master, client->ZK etc.). This prevents HBase from having a single holistic timeout configuration that applies to any RPCs.
+* If there is any issue with ZK (like connection overload / timeouts), the entire HBase service appears frozen and there is little visibility into it.
+* Exposing zookeeper to all the clients can be risky since it can potentially be abused to DDOS.
+* Embedded ZK client is bundled with hbase client jar as a dependency (with it’s log spew :-]).
+
+== Goal
```
Comment (saintstack) [inline dev-support/design-docs/HBASE-18095-Zookeeper-less-client-connection-design.adoc]: s/Goal/Description/

File: dev-support/design-docs/HBASE-18095-Zookeeper-less-client-connection-design.adoc [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+We would like to remove this ZK dependency in the HBase client and instead have the clients query a preconfigured list of active and standby master host:port addresses. This brings all the client interactions with HBase under the same RPC framework that is holistically controlled by a set of hbase client configuration parameters. This also alleviates the pressure on ZK cluster which is critical from an operational standpoint as some core processes like replication, log splitting, master election etc. depend on it. The next section describes the kind of changes needed on both server and client side to support this behavior.
```
Comment (saintstack) [inline dev-support/design-docs/HBASE-18095-Zookeeper-less-client-connection-design.adoc]: s/We would like to remove this/This feature removes the/

File: dev-support/design-docs/HBASE-18095-Zookeeper-less-client-connection-design.adoc [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+== Proposed design
```
Comment (saintstack) [inline dev-support/design-docs/HBASE-18095-Zookeeper-less-client-connection-design.adoc]: This is the refguide. Audience is operators. Also feature is 'done' so s/Proposed//

File: dev-support/design-docs/HBASE-18095-Zookeeper-less-client-connection-design.adoc [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+-----
+
+Clients should be robust enough to handle configuration changes to this parameter since master hosts can change (added/removed) over time and not every client can afford a restart.
+
+One thing to note here is that having masters in the init/read/write path for clients means that
+
+* At Least one active/standby master is now needed for connection creation. Earlier this was not a requirement because the clients looked up the cluster ID from the relevant znode and init successfully. So, technically a master need not be around to create a connection to the cluster.
+* Masters are now active part of read write path in client life cycle under certain scenarios. If the client  cache of meta locations/active master is purged/stale, at least one master (active/stand-by) serving the latest information should exist. Earlier this information was served by ZK and clients look up the latest cluster ID/active master/meta locations from the relevant znodes and get going.
+* There is a higher connection load on the masters than before.
+* More state synchronization traffic (see below)
```
Comment (saintstack) [inline dev-support/design-docs/HBASE-18095-Zookeeper-less-client-connection-design.adoc]: Good

File: dev-support/design-docs/HBASE-18095-Zookeeper-less-client-connection-design.adoc [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+End users should factor these requirements into their cluster deployment if they intend to use this feature.
+
+=== Server side changes
+
+Now that the master end points are considered as source of truth for clients, they should track the latest meta information for clusterID, active master and meta table locations. Since the clients can connect to any master end point (details below), all the masters (active/standby) now track all the relevant meta information. The idea is to implement an in-memory cache local to all the masters and it should keep up with changes to this metadata. This is tracked in the following jiras.
```
Comment (saintstack) [inline dev-support/design-docs/HBASE-18095-Zookeeper-less-client-connection-design.adoc]: s/they should/they/

Comment (saintstack) [reply dev-support/design-docs/HBASE-18095-Zookeeper-less-client-connection-design.adoc]: s/is to implement/is that we implement/

File: dev-support/design-docs/HBASE-18095-Zookeeper-less-client-connection-design.adoc [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+* Clusterid tracking - https://issues.apache.org/jira/browse/HBASE-23257[HBASE-23257]
+* Active master tracking - https://issues.apache.org/jira/browse/HBASE-23275[HBASE-23275]
+* Meta location tracking - https://issues.apache.org/jira/browse/HBASE-23281[HBASE-23281]
```
Comment (saintstack) [inline dev-support/design-docs/HBASE-18095-Zookeeper-less-client-connection-design.adoc]: TMI No need of these JIRA listings.

File: dev-support/design-docs/HBASE-18095-Zookeeper-less-client-connection-design.adoc [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+  <value>master1:16000,master2:16001,master3:16000</value>
+</property>
+-----
+
+Clients should be robust enough to handle configuration changes to this parameter since master hosts can change (added/removed) over time and not every client can afford a restart.
+
+One thing to note here is that having masters in the init/read/write path for clients means that
+
+* At Least one active/standby master is now needed for connection creation. Earlier this was not a requirement because the clients looked up the cluster ID from the relevant znode and init successfully. So, technically a master need not be around to create a connection to the cluster.
+* Masters are now active part of read write path in client life cycle under certain scenarios. If the client  cache of meta locations/active master is purged/stale, at least one master (active/stand-by) serving the latest information should exist. Earlier this information was served by ZK and clients look up the latest cluster ID/active master/meta locations from the relevant znodes and get going.
```
Comment (ndimiduk) [inline dev-support/design-docs/HBASE-18095-Zookeeper-less-client-connection-design.adoc]: I know we nit-picked over this language earlier. With this feature enabled, this change places the masters into the client lifecycle ~under _all_ scenarios~. It's simply the lifecycle. That involvement mirrors the current ZK involvement, which is to say that the masters are introduced into the path of service discovery that a client undertakes in order to locate the data it seeks.

File: dev-support/design-docs/HBASE-18095-Zookeeper-less-client-connection-design.adoc [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+=== Client side changes
+
+The proposal is to implement a new AsyncRegistry (https://issues.apache.org/jira/browse/HBASE-23305[HBASE-23305]) based on the master RPC endpoints discussed above. Few interesting optimizations on the client side as follows.
+
+* Each client randomly picks a master from the list of host:ports passed in the configuration. This avoids hotspotting on a single master.
+* Client can also do hedged lookups, meaning a single RPC to fetch meta information (say active master) can be sent to multiple masters and which ever returns first can be passed along to the caller. This can be done under a config flag since it comes with an additional RPC load. The default behavior is to randomly probe masters until the list is exhausted.
+* Callers are expected to cache the meta information in higher levels and only probe again if the cached information is stale (which is quite possible).
+
+One user noted that there are some clients that rely on cluster ID for delegation token based auth. So there is a proposal to expose it on an auth-less end point for delegation auth support..
```
Comment (ndimiduk) [inline dev-support/design-docs/HBASE-18095-Zookeeper-less-client-connection-design.adoc]: Is there an issue filed for this? Might as well link it here.

File: src/main/asciidoc/_chapters/architecture.adoc [hunk 9]
```diff
@@ -260,6 +260,73 @@ For region name, we only accept `byte[]` as the parameter type and it may be a f
 
 Information on non-Java clients and custom protocols is covered in <<external_apis>>
 
+[[client.masterregistry]]
+=== Master registry (new as of release 3.0.0)
```
Comment (ndimiduk) [inline src/main/asciidoc/_chapters/architecture.adoc]: New as of 2.3.0, right? I think you'll want to update this throughout the doc.

File: src/main/asciidoc/_chapters/architecture.adoc [hunk 10]
```diff
@@ -260,6 +260,73 @@ For region name, we only accept `byte[]` as the parameter type and it may be a f
...
+adapt to any sort of live rpc performance metrics.
+
+==== Additional Notes
+
+* Clients hedge the requests in a randomized order to avoid hot-spotting a single server.
+* Cluster internal connections (master<->regionservers) still use ZooKeeper based connection
+registry.
+* Cluster internal state is still tracked in Zookeeper, hence ZK availability requirements are same
+as before.
+* Inter cluster replication still uses ZooKeeper beased connection registry to simplify configuration
```
Comment (ndimiduk) [inline src/main/asciidoc/_chapters/architecture.adoc]: s/beased/based/

File: src/main/asciidoc/_chapters/architecture.adoc [hunk 11]
```diff
@@ -260,6 +260,73 @@ For region name, we only accept `byte[]` as the parameter type and it may be a f
...
+management.
+
+For more implementation details, please refer to the https://github.com/apache/hbase/tree/master/dev-support/design-docs[design doc] and
+https://issues.apache.org/jira/browse/HBASE-18095[HBASE-18095].
+
+'''
+NOTE: (Advanced) In case of any issues with the master based registry, use the following
+configuration to fallback to the ZooKeeper based connection registry implementation.
```
Comment (ndimiduk) [inline src/main/asciidoc/_chapters/architecture.adoc]: It's good that this is gated by a single config change.

File: src/main/asciidoc/_chapters/configuration.adoc [hunk 12]
```diff
@@ -563,38 +563,63 @@ Changes here will require a cluster restart for HBase to notice the change thoug
...
-Usually this ensemble location is kept out in the _hbase-site.xml_ and is picked up by the client from the `CLASSPATH`.
+Starting release 3.0.0, the default connection registry has been switched to a master based implementation. Refer to <<client.masterregistry>> for more details about
+what a connection registry is and implications of this change. Depending on your HBase version, following is the expected minimal client configuration.
 
-If you are configuring an IDE to run an HBase client, you should include the _conf/_ directory on your classpath so _hbase-site.xml_ settings can be found (or add _src/test/resources_ to pick up the hbase-site.xml used by tests).
+==== Up until 2.x.y releases
+In 2.x.y releases, the default connection registry was based on ZooKeeper as the source of truth. This means that the clients always looked up ZooKeeper znodes to fetch
+the required metadata. For example, if an active master crashed and the a new master is elected, clients looked up the master znode to fetch
+the active master address (similarly for meta locations). This meant that the clients needed to have access to ZooKeeper and need to now
```
Comment (ndimiduk) [inline src/main/asciidoc/_chapters/configuration.adoc]: a/now/know/

File: src/main/asciidoc/_chapters/configuration.adoc [hunk 13]
```diff
@@ -563,38 +563,63 @@ Changes here will require a cluster restart for HBase to notice the change thoug
...
 ----
 
+The configuration value for _hbase.masters_ is a comma separated list of _host:port_ values. If no port value is specified, the default of _16000_ is assumed.
+
+Usually this configuration is kept out in the _hbase-site.xml_ and is picked up by the client from the `CLASSPATH`.
+
+If you are configuring an IDE to run an HBase client, you should include the _conf/_ directory on your classpath so _hbase-site.xml_ settings can be found (or add _src/test/resources_ to pick up the hbase-site.xml used by tests).
+
+For Java applications using Maven, including the hbase-shaded-client module is the recommended dependency when connecting to a cluster:
+[source,xml]
```
Comment (ndimiduk) [inline src/main/asciidoc/_chapters/configuration.adoc]: nit: is dependency information relevant in the configuration section?

**Comment by ndimiduk** [comment]
Oh, I see you merged the PR already. The JIRA was still open, so I assumed it was not in yet.

File: dev-support/design-docs/HBASE-18095-Zookeeper-less-client-connection-design.adoc (hunk 8 above)
Comment (bharathv) [reply dev-support/design-docs/HBASE-18095-Zookeeper-less-client-connection-design.adoc]: HBASE-23330, yes that was fixed too.

File: src/main/asciidoc/_chapters/architecture.adoc (hunk 9 above)
Comment (bharathv) [reply src/main/asciidoc/_chapters/architecture.adoc]: I wanted to update the docs once it actually makes it to the 2.3.0...that ok? (didn't want to do that without committing to the branch).

File: src/main/asciidoc/_chapters/architecture.adoc (hunk 10 above)
Comment (bharathv) [reply src/main/asciidoc/_chapters/architecture.adoc]: Done.

File: src/main/asciidoc/_chapters/architecture.adoc (hunk 11 above)
Comment (bharathv) [reply src/main/asciidoc/_chapters/architecture.adoc]: Yep.

File: src/main/asciidoc/_chapters/configuration.adoc (hunk 12 above)
Comment (bharathv) [reply src/main/asciidoc/_chapters/configuration.adoc]: Done.

File: src/main/asciidoc/_chapters/configuration.adoc (hunk 13 above)
Comment (bharathv) [reply src/main/asciidoc/_chapters/configuration.adoc]: ya seems out of place but the section is named "Client configuration and dependencies connecting to an HBase cluster".. so probably was intentional.

File: dev-support/design-docs/HBASE-18095-Zookeeper-less-client-connection-design.adoc (hunk 7 above)
Comment (bharathv) [reply dev-support/design-docs/HBASE-18095-Zookeeper-less-client-connection-design.adoc]: Fixed.

**Pull request merged**